src/commons/eventsourcing/store.rs

```diff
@@ -178,152 +92,94 @@ where
+                Ok(init_command) => {
+                    if let Some(init_event) = init_command.into_init() {
+                        // Initialise
+                        let mut aggregate = A::init(handle.clone(), init_event);
+
+                        // Find the next command and apply it until there
+                        // is no (valid) next command.
+                        while let Ok(command) = self.get_command(&handle, aggregate.version()) {
+                            aggregate.apply_command(command);
                         }
-                        last_good_cmd = cmd.sequence();
-                        last_update = cmd.time();
+
+                        // Ret
+                        Some(aggregate)
                     } else {
-                        all_ok = false;
+                        None
                     }
                 }
-                if !all_ok {
+            };
+
+            match aggregate_opt {
+                None => {
                     warn!(
-                        "Command {} was corrupt, or not all events could be loaded. Will archive surplus",
-                        command_key
+                        "No valid initialisation command found for '{}', will remove it.",
+                        handle
                     );
-                    // Bad command or event encountered.. archive surplus commands
-                    // note that we will clean surplus events later
-                    self.archive_surplus_command(&handle, &command_key)?;
-                }
-            }
-
-            self.archive_surplus_events(&handle, last_good_evt + 1)?;
 
-            if !all_ok {
-                warn!(
-                    "State for '{}' can only be recovered to version: {}. Check corrupt and surplus dirs",
-                    &handle, last_good_evt
-                );
-            }
-
-            // Get the latest aggregate, not that this ensures that the snapshots
-            // are checked, and archived if corrupt, or if they are after the last_good_evt
-            let agg = self
-                .get_aggregate(&handle, Some(last_good_evt))?
-                .ok_or_else(|| AggregateStoreError::CouldNotRecover(handle.clone()))?;
-
-            let snapshot_version = agg.version();
-
-            let info = StoredValueInfo {
-                last_event: last_good_evt,
-                last_command: last_good_cmd,
-                last_update,
-                snapshot_version,
-            };
+                    self.kv.drop_scope(&scope)?;
+                }
+                Some(aggregate) => {
+                    // Archive any and all keys that are not command keys
+                    // for versions we just applied.
+                    for key in self.kv.keys(&scope, "")? {
+                        // command keys use: command-#.json
+                        let keep = if let Some(pfx_removed) = key.name().as_str().strip_prefix("command-") {
```

Comment:
You could simplify this with a `let mut keep = false` and only set it to `true` in the deepest if level.